### ISSUE TITLE
Skinned mesh character gets distorted after conversion

### ISSUE THREAD

**Issue by fedemarino31** [KhronosGroup/COLLADA2GLTF#95]
Hi everyone, I'm trying to convert an animated skinned charater to use it in Three.js
The DAE works great but the GLTF appears distorted as if the mesh has some transformation prior to the binding of the Skin.

[
![gltfproblem](https://user-images.githubusercontent.com/13981045/30442799-e933ee0a-9953-11e7-8f27-a1dab7bc1214.jpg)
](url)

Here you have the 2 files (original and converted)
[models.zip](https://github.com/KhronosGroup/COLLADA2GLTF/files/1303776/models.zip)

**Comment by javagl** [comment]
Very likely a similar issue as https://github.com/KhronosGroup/glTF-Sample-Models/issues/83 . 

@lasalvavida I'm just trying to investigate this further in COLLADA2GLTF, but ... can't promise anything. However, a code search on GitHub looks like the `bindShapeMatrix` is not used at all. (There is a setter and a getter, but no one calling the latter). Considering that this model **also** has a `bindShapeMatrix` that is **not** the identity matrix, it seems very likely that this is the reason....

**Comment by lasalvavida** [comment]
@javagl, any help you're willing to provide is always appreciated!

**Comment by javagl** [comment]
As I guessed in my [long comment in the other issue](https://github.com/KhronosGroup/glTF-Sample-Models/issues/83#issuecomment-329541086), the `inverseBindMatrices` are erreonously premultiplied with the `bindShapeMatrix`.

*This* part of the fix is trivial: Just comment out https://github.com/KhronosGroup/COLLADA2GLTF/blob/be4338c9f510a062dd87e39178b6772d335e5fc9/src/COLLADA2GLTFWriter.cpp#L1437

The problem now is that the `bindShapeMatrix` has to be multiplied into the `POSITIONS`. I thought that this might happen somewhere around https://github.com/KhronosGroup/COLLADA2GLTF/blob/be4338c9f510a062dd87e39178b6772d335e5fc9/src/COLLADA2GLTFWriter.cpp#L655 , with something like (pseudocode) 

    if(semantic=="POSITION") {
        vec3 p = buildAttributes[semantic][last];
        buildAttributes[semantic][last] = transform(bindShapeMatrix, p);
    }

but there are several difficulties (for me, not being involved in the code and COLLADA in general) : The `bindShapeMatrix` is not known there, because it is defined in the node ("controller"?) that refers to the mesh. Additionally, the mesh might be referred to by *multiple* nodes, with different bind shape matrices - it might be necessary to "instantiate" the mesh (or its primitive) several times. 

In any case, if all this is true, it looks like (possibly considerable) refactoring. I could create a PR for the `//`, but not for this ;-)

**Comment by lasalvavida** [comment]
It actually gets a little tricker than that. COLLADA files are read as an XML stream (SAX). Very likely the skinning information hasn't been loaded yet when we are creating the mesh and primitives.

Couldn't we just insert a node with the bindShapeMatrix transform (if it is non-identity) to contain the skinned mesh instead of editing the data (we try to avoid this because it is slow).

**Comment by javagl** [comment]
Indeed, fortunately, it may even be much simpler. After having another look at the spec, https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#skins it says:

> Implementation Note: Matrix, defining how to pose the skin's geometry for use with the joints ("Bind Shape Matrix") should be premultiplied to mesh data or to Inverse Bind Matrices.

and considering that the joint matrices are just computed using 

        // [jointMatrix(j)] = 
        //     [globalTransformOfNode^-1] *
        //     [globalTransformOfJointNode] *
        //     [inverseBindMatrix(j)] *
        //     [bindShapeMatrix]

it is possible to inline the `bindShapeMatrix` into the joint matrix (and thus, avoid having to modify the `POSITIONS` and the hassle that could result from that). 

So from what I understood, it should already be solved by not **pre**-multiplying the `bindShapeMatrix` with the `inverseBindMatrix`, but instead, **post**-multiplying it. 

I tested it with the "WalkingLady" model, and it *appears* to be correct. 

(I only tested it with my local JglTF library, so whether or not it is correct should still be confirmed by others) 

If it works, here is a PR https://github.com/KhronosGroup/COLLADA2GLTF/pull/96
